core/src/main/java/org/apache/carbondata/core/datamap/DistributableDataMapFormat.java

```diff
@@ -16,62 +16,115 @@
 import org.apache.hadoop.mapreduce.RecordReader;
 import org.apache.hadoop.mapreduce.TaskAttemptContext;
 import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.log4j.Logger;
 
 /**
  * Input format for datamaps, it makes the datamap pruning distributable.
  */
-public class DistributableDataMapFormat extends FileInputFormat<Void, ExtendedBlocklet> implements
-    Serializable {
+public class DistributableDataMapFormat extends FileInputFormat<Void, ExtendedBlocklet>
+    implements Serializable, Writable {
+
+  private static final transient Logger LOGGER =
+      LogServiceFactory.getLogService(DistributableDataMapFormat.class.getName());
+
+  private static final long serialVersionUID = 9189779090091151248L;
 
   private CarbonTable table;
 
+  private FilterResolverIntf filterResolverIntf;
+
   private DataMapExprWrapper dataMapExprWrapper;
 
   private List<Segment> validSegments;
 
-  private List<Segment> invalidSegments;
+  private List<String> invalidSegments;
 
   private List<PartitionSpec> partitions;
 
-  private  DataMapDistributableWrapper distributable;
-
   private boolean isJobToClearDataMaps = false;
 
-  DistributableDataMapFormat(CarbonTable table, DataMapExprWrapper dataMapExprWrapper,
-      List<Segment> validSegments, List<Segment> invalidSegments, List<PartitionSpec> partitions,
-      boolean isJobToClearDataMaps) {
+  private ReadCommittedScope readCommittedScope;
+
+  private DataMapLevel dataMapLevel;
+
+  private boolean isFallbackJob = false;
+
+  DistributableDataMapFormat() {
+
+  }
+
+  DistributableDataMapFormat(CarbonTable table,
+      List<Segment> validSegments, List<String> invalidSegments, boolean isJobToClearDataMaps,
+      DataMapLevel dataMapLevel) throws IOException {
+    this(table, null, validSegments, invalidSegments, null,
+        isJobToClearDataMaps, dataMapLevel, false);
+  }
+
+  DistributableDataMapFormat(CarbonTable table, FilterResolverIntf filterResolverIntf,
+      List<Segment> validSegments, List<String> invalidSegments, List<PartitionSpec> partitions,
+      boolean isJobToClearDataMaps, DataMapLevel dataMapLevel, boolean isFallbackJob)
+      throws IOException {
     this.table = table;
-    this.dataMapExprWrapper = dataMapExprWrapper;
+    this.filterResolverIntf = filterResolverIntf;
     this.validSegments = validSegments;
     this.invalidSegments = invalidSegments;
     this.partitions = partitions;
+    if (!validSegments.isEmpty()) {
+      this.readCommittedScope = validSegments.get(0).getReadCommittedScope();
+    }
     this.isJobToClearDataMaps = isJobToClearDataMaps;
+    this.dataMapLevel = dataMapLevel;
+    this.isFallbackJob = isFallbackJob;
+    generateDataMapExpr();
+  }
+
+  private void generateDataMapExpr() throws IOException {
+    if (dataMapLevel == null) {
+      this.dataMapExprWrapper = DataMapChooser.getDefaultDataMap(table, filterResolverIntf);
```

Comment:
I think there is no need to create dataMapExprWrapper here and serialize it. Better create in the executor side.

Reply:
removed